collection.go

```diff
@@ -246,22 +246,58 @@ func (c *Collection) AddDocument(ctx context.Context, doc Document) error {
 	return nil
 }
 
-// RemoveDocument removes a document from the collection.
-func (c *Collection) RemoveDocument(_ context.Context, documentID string) error {
-	if documentID == "" {
-		return errors.New("documentID is empty")
+// Delete removes document(s) from the collection.
+//
+//   - where: Conditional filtering on metadata. Optional.
+//   - whereDocument: Conditional filtering on documents. Optional.
+//   - ids: The ids of the documents to delete. If empty, all documents are deleted.
+func (c *Collection) Delete(_ context.Context, where, whereDocument map[string]string, ids ...string) error {
+
+	// must have at least one of where, whereDocument or ids
+	if len(where) == 0 && len(whereDocument) == 0 && len(ids) == 0 {
+		return fmt.Errorf("must have at least one of where, whereDocument or ids")
+	}
+
+	if len(c.documents) == 0 {
```

Comment:
As concurrent processes could be writing to the documents, I think we need a lock here (and also around the `filterDocs()` call). There might be three ways, with different tradeoffs re performance, consistency and simplicity:

1. We could either add an `RLock()` call above, and then an `RUnlock()` behind the `filterDocs()`, before then doing the `Lock()` you already put in place below
2. Or have a `Lock()` from above until the end
3. Or lock only exactly where a concurrent operation could get in the way (one here, one around the filterDocs, one where you already have it)

- 1+3 have the risk that the collection can change from a concurrent write, between our doc filtering and deletion.
  - It can be a doc addition, which could leave a doc remaining while it might fit a filter that was passed to the deletion.
  -  It can be a doc deletion, and then for the persistent store we'll run into "file not found" or similar error.
- 2 has the downside that the write lock is an exclusive lock, so during the entire deletion process the collection can't be used for other purposes. But it's more consistent

It depends a lot on the application that uses `chromem-go` which of those is better I guess. My use case so far was a fairly strong split between filling the DB, and querying it, so there's no issue with 1+3. But some people might use it in a web service with multiple users, and then concurrent requests with doc additions + query + deletions are more likely to happen.

WDYT?

If we go with the third option, then here we can use:

```suggestion
	if c.Count() == 0 {
```

Reply:
For simplicity, and assuming (at least for now) that the library is not used for huge collections where the lock duration during deletion is an issue, I think I would lean towards just locking from earlier to the end.

But I'm open to discussing it further. In any case we can also change it in the future as it's internal and doesn't affect the signature.

Reply:
Oh yeah, I missed that, sorry. Didn't think about this, while rewriting this "on the road" - Updated to just lock before filtering. :+1:

Reply:
Ah right it's enough to move it before the `filterDocs()`, no need here for the `len`, as that reads the len field from the underlying structure atomically and is not a regular slice read operation :ok_hand: